internal/pkg/cli/delete/static_site.go

```diff
@@ -0,0 +1,44 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package delete
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/aws/copilot-cli/internal/pkg/s3"
+)
+
+type bucketResourceGetter interface {
+	BucketName(app, env, svc string) (string, error)
+}
+
+type bucketEmptier interface {
+	EmptyBucket(string) error
+}
+
+// StaticSiteDeleter is used to clean up resources created for a static site.
+type StaticSiteDeleter struct {
+	BucketResourceGetter bucketResourceGetter
+	BucketEmptier        bucketEmptier
```

Comment:
I agree with your reasons listed above for this case, but want to probe a little further: what about structs that have, say, 7 **required** fields? Beyond 5 or 6 function parameters, it's pretty normal to pass a struct for function parameters. So is this preferred to just saying all fields on `Foo` are required?

```
type Foo struct {
  // 7 required fields
}

type FooRequired struct {
  // 7 required fields
}

func New(req FooRequired) Foo {
  return Foo{
    // apply req
  }
}
```

and how about if there were additional optional fields? something like this?
```
type Foo struct {
  // 7 required fields
  // 3 optional fields
}

type FooRequired struct {
  // 7 required fields
}

type FooOption func(*Foo)

func WithOptionA(a string) FooOption {
  return func(f *Foo) {
    f.a = a
  }
}

func NewFoo(req FooRequired, opts ...FooOption) Foo {
  f := Foo{
    //apply req
  }
  
  // apply options
  return f
}
```

In both of these cases, the alternative approach in my mind is just labelling the fields, like:
```
type Foo struct {
  // A is required
  A string
  // 6 more required

  // X is optional
  X string
  // 2 more optional
}
```

I see benefits/drawbacks both ways, so I'm just curious on your thoughts?

Reply:
ugh honestly I think the last option: 
```
type Foo struct {
  // A is required
  A string
  // 6 more required

  // X is optional
  X string
  // 2 more optional
}
```
Is only OK if actually all the fields are _optional_, i.e. the struct internally can call an [`init` method](https://github.com/robpike/ivy/blob/f066ab6d0b3dd95e6cc5fce39ab5f2a3ffd6f9a0/config/config.go#L76) and set the values. Otherwise, I'd avoid this option.

My preferred option is actually like this when there are many parameters:
```go
type Foo struct {
  // 7 required fields
  // 3 optional fields
}

type FooConfig struct {
  // 7 required fields
}

type FooOpts struct {
  // 3 optional fields.
}

func NewFoo(cfg FooConfig) Foo
func NewFooWithOpts(cfg FooConfig, opts FooOpts) Foo
```